File: guests/haskell/panes.hs
{-# LANGUAGE OverloadedStrings #-}

-- The panes scene, Haskell port — guests/rust/panes.rs,
-- tools/scenes/panes.steps.

import Data.Word (Word64)
import Data.Text (Text)
import qualified Data.Text as T
import KayaApp

contentId, detailId :: Word64
contentId = 7
detailId = 8

main :: IO ()
main = kayaMain $ \app -> do
  buildTx app $ do
    window primary [WTitle "panes", WPanes 3]
    caption <- signal (T.pack "root pane")
    root <-
      column
        []
        [ -- Authored ids so the REAL-TREE read can address these: an
          -- index read passes whether or not anything reached the
          -- screen.
          labelBound caption [A11yId ("root" :: Text)], -- label#0
          buttonOn
            "open content"
            ( buildTx app $ do
                pushEntry contentId [ETitle "content"]
                inner <- signal (T.pack "content pane")
                pane <-
                  column
                    []
                    [ labelBound inner [A11yId ("content" :: Text)], -- label#1
                      buttonOn
                        "open detail"
                        ( buildTx app $ do
                            pushEntry detailId [ETitle "detail"]
                            leaf <- signal (T.pack "detail pane")
                            deep <- column [] [labelBound leaf [A11yId ("detail" :: Text)]] -- label#last
                            mountIn detailId deep
                        )
                        [] -- button#1
                    ]
                mountIn contentId pane
            )
            [] -- button#0
        ]
    mount root

File: guests/rust/panes.rs
//! The panes scene (tools/scenes/panes.steps): nothing here is
//! panes-specific except `panes(3)`, asked for ONCE.

pub(crate) fn app(ctx: kaya::AppCtx) {
    use kaya::WindowId;

    #[derive(Clone, Copy)]
    enum Msg {
        OpenContent,
        OpenDetail,
    }

    const CONTENT: WindowId = WindowId(7);
    const DETAIL: WindowId = WindowId(8);

    let msgs = kaya::Messages::<Msg>::new();
    ctx.apply(|tx| {
        tx.window(kaya::DEFAULT_WINDOW).title("panes").panes(3);
        let root = tx
            .column(|tx| {
                let caption = tx.signal("root pane");
                tx.label(caption).a11y_id("root"); // label#0
                let open = tx.button("open content").id(); // button#0
                msgs.on_click(open, Msg::OpenContent);
            })
            .id();
        tx.mount(root);
    });

    while let Some(msg) = msgs.next(&ctx) {
        match msg {
            Msg::OpenContent => ctx.apply(|tx| {
                let entry = tx.push_entry(CONTENT).title("content").id();
                let pane = tx
                    .column(|tx| {
                        let caption = tx.signal("content pane");
                        tx.label(caption).a11y_id("content"); // label#1
                        let open = tx.button("open detail").id(); // button#1
                        msgs.on_click(open, Msg::OpenDetail);
                    })
                    .id();
                tx.mount_in(entry, pane);
            }),
            Msg::OpenDetail => ctx.apply(|tx| {
                let entry = tx.push_entry(DETAIL).title("detail").id();
                let pane = tx
                    .column(|tx| {
                        let caption = tx.signal("detail pane");
                        tx.label(caption).a11y_id("detail"); // label#last
                    })
                    .id();
                tx.mount_in(entry, pane);
            }),
        }
    }
}

fn main() {
    kaya::run(app)
}
